File: src/apps/Andriod/app/src/main/java/com/dkellycollins/triggerfinger/managers/view/impl/BaseSpriteViewManager.java
package com.dkellycollins.triggerfinger.managers.view.impl;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.RectF;

import com.dkellycollins.triggerfinger.data.entity.ICollidable;
import com.dkellycollins.triggerfinger.managers.entity.IBitmapEntityManager;
import com.dkellycollins.triggerfinger.managers.view.IViewManager;

public abstract class BaseSpriteViewManager implements IViewManager {

    private final IBitmapEntityManager _bitmapManager;

    protected BaseSpriteViewManager(IBitmapEntityManager bitmapManager) {
        _bitmapManager = bitmapManager;
    }

    @Override
    public void init() {
        _bitmapManager.load(getSpriteId());
    }

    @Override
    public void dispose() {
        _bitmapManager.delete(getSpriteId());
    }

    protected void drawSprite(Canvas canvas, ICollidable collidable) {
        Bitmap sprite = _bitmapManager.retrieve(getSpriteId());

        canvas.drawBitmap(sprite, null, getDestinationRect(collidable), null);
    }


    protected abstract int getSpriteId();

    private RectF getDestinationRect(ICollidable collidable) {
        float top = collidable.getCenter().getY() - collidable.getRadius();
        float left = collidable.getCenter().getX() - collidable.getRadius();
        float bottom = collidable.getCenter().getY() + collidable.getRadius();
        float right = collidable.getCenter().getX() + collidable.getRadius();

        return new RectF(left, top, right, bottom);
    }
}

File: src/apps/Android-Kotlin/app/src/main/kotlin/com/dkellycollins/triggerfinger/managers/view/impl/BaseSpriteViewManager.kt
package com.dkellycollins.triggerfinger.managers.view.impl

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.RectF

import com.dkellycollins.triggerfinger.data.entity.ICollidable
import com.dkellycollins.triggerfinger.managers.entity.IBitmapEntityManager
import com.dkellycollins.triggerfinger.managers.view.IViewManager

abstract class BaseSpriteViewManager protected constructor(private val _bitmapManager: IBitmapEntityManager) : IViewManager {

    override fun init() {
        _bitmapManager.load(spriteId)
    }

    override fun dispose() {
        _bitmapManager.delete(spriteId)
    }

    protected fun drawSprite(canvas: Canvas, collidable: ICollidable) {
        val sprite = _bitmapManager.retrieve(spriteId)

        canvas.drawBitmap(sprite, null, getDestinationRect(collidable), null)
    }


    protected abstract val spriteId: Int

    private fun getDestinationRect(collidable: ICollidable): RectF {
        val top = collidable.center.y - collidable.radius
        val left = collidable.center.x - collidable.radius
        val bottom = collidable.center.y + collidable.radius
        val right = collidable.center.x + collidable.radius

        return RectF(left, top, right, bottom)
    }
}
